src/hotspot/share/runtime/lockStack.inline.hpp

```diff
@@ -104,16 +104,11 @@ inline void LockStack::remove(oop o) {
 
 inline bool LockStack::contains(oop o) const {
   verify("pre-contains");
-  if (!SafepointSynchronize::is_at_safepoint() && !is_owning_thread()) {
-    // When a foreign thread inspects this thread's lock-stack, it may see
-    // bad references here when a concurrent collector has not gotten
-    // to processing the lock-stack, yet. Call StackWaterMark::start_processing()
-    // to ensure that all references are valid.
-    StackWatermark* watermark = StackWatermarkSet::get(get_thread(), StackWatermarkKind::gc);
-    if (watermark != nullptr) {
-      watermark->start_processing();
-    }
-  }
+
+  // Can't poke around in thread oops without having started stack watermark processing.
+  StackWatermark* watermark = StackWatermarkSet::get(get_thread(), StackWatermarkKind::gc);
+  assert(watermark == nullptr || watermark->processing_started(), "Processing must have started!");
```

Comment:
Is line #109 only to support the assertion? If so it should be ifdef'd. Though for an assertion I would prefer to see some kind of `StackWatermark::is safe_for(Thread* t)` API.

Reply:
Yes, it is only for the assertions. I prefer to avoid adding ifdefs because I find them highly distracting. I was considering adding a function, but couldn't immediately come up with a great name. I'll send out a new proposal with a new function. Thanks.

Reply:
We already had a `StackWatermarkSet::processing_started(JavaThread*)` function, so I hanged the code to use it.